src/lib/comms/sol-lwm2m.c

```diff
@@ -4718,6 +4859,172 @@ client_bootstrap(void *data)
+    int r;
+    uint16_t i;
+    struct obj_instance *instance;
+    struct sol_lwm2m_resource res;
+    struct sol_vector server_ids, acl_values;
+    uint16_t *server_id;
+    int64_t *acl_value;
+
+    sol_vector_init(&server_ids, sizeof(uint16_t));
+    sol_vector_init(&acl_values, sizeof(int64_t));
+
+    obj_ctx = find_object_ctx_by_id(client, SERVER_OBJECT_ID);
+    if (obj_ctx && obj_ctx->instances.len > 0) {
+        //If any Server Object Instance exists, this is already Bootstrapped and
+        // one ACL will be created per existing [Server ID]
+        SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, instance, i) {
+            r = read_resources(client, obj_ctx, instance, &res, 1,
+                SERVER_OBJECT_SERVER_ID);
+            if (r < 0) {
+                SOL_WRN("Could not read Server Object's [Server ID] resource\n");
+                return r;
+            }
+            server_id = sol_vector_append(&server_ids);
+            *server_id = res.data->content.integer;
+
+            acl_value = sol_vector_append(&acl_values);
+            *acl_value = SOL_LWM2M_ACL_CREATE;
+
+            sol_lwm2m_resource_clear(&res);
+        }
+    } else {
+        //Else, there's no way to know which Server is authorized to
+        // instantiate which Object, and thus <no Server> should be authorized
+        // to instantiate <any> Object. However, since there's no way for the Bootstrap
+        // Server to Read /2, in order to get the Access Control Instance relevant for
+        // 'C'reate on a given Object, and change the ACLs granting access rights to a given
+        // Server, <every Server> will be authorized to instantiate <every> Object
+        SOL_DBG("LWM2M Server object not provided! No factory bootstrap.");
+
+        server_id = sol_vector_append(&server_ids);
+        *server_id = DEFAULT_SHORT_SERVER_ID;
+
+        acl_value = sol_vector_append(&acl_values);
+        *acl_value = SOL_LWM2M_ACL_CREATE;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&client->objects, obj_ctx, i) {
+        //No one is allowed to 'C'reate Security Objects, Server Objects,
+        // and Access Control Objects
+        if (obj_ctx->obj->id != SECURITY_SERVER_OBJECT_ID &&
+            obj_ctx->obj->id != SERVER_OBJECT_ID &&
+            obj_ctx->obj->id != ACCESS_CONTROL_OBJECT_ID) {
+            SOL_SET_API_VERSION(res.api_version = SOL_LWM2M_RESOURCE_API_VERSION; )
+            r = sol_lwm2m_resource_init_vector(&res, ACCESS_CONTROL_OBJECT_ACL_RES_ID,
+                SOL_LWM2M_RESOURCE_DATA_TYPE_INT, &server_ids, &acl_values);
+            if (r < 0) {
+                SOL_WRN("Could not init Access Control Object's [ACL] resource\n");
+                return r;
+            }
+
+            r = setup_access_control_object_instance_for_instance(client,
+                obj_ctx->obj->id, UINT16_MAX, UINT16_MAX, &res, false);
+
+            sol_vector_clear(&server_ids);
+            sol_vector_clear(&acl_values);
```

Comment:
Wow, thanks.

Moved these operations to outside the loop (and added them to the failing `else` as well).